components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -288,6 +288,20 @@ open class GleanInternalAPI internal constructor () {
         return ExperimentsStorageEngine.getSnapshot().getValue(experimentId)
     }
 
+    /**
+     * Provides a way to clear all stores and serialized pings from a test, and set Glean back to an
+     * uninitialized state. When outside libraries use the testAPI, it may be difficult to reset the
+     * metrics and Glean to provide a blank slate to test from.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    fun testResetGlean() {
```

Comment:
What is the problem you're trying to solve here? Which test is failing?

```
// Start of the test
val initialValue = theMetric.testGetValue()

// Do something test-specific

val finalValue = testMetric.testGetValue()

// Check that it stored the right thing
assertEqual(expectedValue, finalValue - initialValue)
```

Could the above work? I'd really try to hold off from this change for now, as this requires a broader discussion on how to do that (= proposal). Maybe we don't want to expose an API like that, maybe we just want to expose a JUnit rule that does that automatically.

Reply:
The problem is that `Glean` singleton is maintaining state between tests.  I can run each individual test by itself and not see this, but if I run the entire suite of tests for this module then it maintains some state between tests within the `Glean` singleton.  This is throwing off the clean state from which to start a test from.

I considered just measuring the change from initial value, but unfortunately there is a test in there that checks that experiments does not crash if Glean is not initialized, and unless you know a way to control the order that the tests are run in, I cannot guarantee that test gets ran prior to any test that initializes Glean.  That is significant because there is currently no way to _de-init_ Glean.

Reply:
> 
> 
> The problem is that `Glean` singleton is maintaining state between tests. I can run each individual test by itself and not see this, but if I run the entire suite of tests for this module then it maintains some state between tests within the `Glean` singleton. This is throwing off the clean state from which to start a test from.

I mean, what's the specific test that's failing when you run the full suite of tests?

> I considered just measuring the change from initial value, but unfortunately there is a test in there that checks that experiments does not crash if Glean is not initialized, and unless you know a way to control the order that the tests are run in, I cannot guarantee that test gets ran prior to any test that initializes Glean. That is significant because there is currently no way to _de-init_ Glean.

You would need to get the initial value within each test, and check at the end of each test. That should work, unless I'm missing something here.

Reply:
Basically every test that runs after the first test that initializes Glean and relies on a value from the Glean testAPI.  In this case, it's 3 tests.  If I rename them so that the order of the tests changes, then the specific tests that fail may change.

I _still need_ to have a way to run a test _before_ Glean is initialized, to guarantee that interacting with an experiment when Glean is not initialized does not crash.  Unless you know of a way to make the tests run in a particular order every time, I cannot guarantee that test gets ran before a test that initializes Glean.  Once Glean is initialized in the context of testing the whole suite of tests, there is no way to get `Glean.isInitialized()` to return false, so I cannot guarantee that Glean is not initialized for the test that it _has_ to be uninitialized for because I cannot control the order that the tests are run in.

Reply:
On top of that, having interdependent tests just seems _wrong_...